Deprecate `ItemModifier` and `ItemDecorator` traits

aochagavia opened the pull request:
Breaking changes:
* Deprecate `ItemModifier` trait (use `MultiItemModifier` instead)
* Deprecate `ItemDecorator` trait (use `MultiItemDecorator` instead)
* Deprecate `SyntaxExtension::Decorator` and `SyntaxExtension::Modifier` (use the `MultiDecorator` and `MultiModifier` variants instead)

[breaking-change]

Fixes #21155 and #21154

rust-highfive commented:
r? @brson

(rust_highfive has picked a reviewer for you, use r? to override)

aochagavia commented:
@nick29581 Is this ok? Do you have any suggestions?

brson commented:
r? @nick29581

nick29581 commented:
LGTM. It needs a test - I think there is probably already a test for ItemDecorator somewhere, you could change that to use MultiItemDecorator, or you could create a new test for the Multi- version.

aochagavia commented:
What kind of test should it be? run-pass?

I was unable to find a test for `ItemDecorator`. Could you point me to it?

nick29581 commented:
probably run-pass-fulldeps. It doesn't look like there is a test for ItemDecorator, you should copy the one for Modifier/MultiModifier. Start with `auxiliary/macro_crate_test.rs` and see where the macros get used (not sure off the top of my head).

aochagavia commented:
@nick29581 I have added a test

nick29581 commented:
The test looks great, a few comments on the other commit though

aochagavia commented:
@nick29581 I have addressed your comments. Thanks for reviewing!

I wasn't sure what to do to with `cfg_attr`. Right now it can be applied on any annotatable. Is that OK?

aochagavia commented on src/libsyntax/ext/cfg_attr.rs:
```diff
@@ -8,27 +8,70 @@
     if attr::cfg_matches(&cx.parse_sess.span_diagnostic, cx.cfg.as_slice(), &**cfg) {
-        out.attrs.push(cx.attribute(attr.span, attr.clone()));
+        let attr = cx.attribute(attr.span, attr.clone());
+        fold_annotatable!(ann, item => item.attrs.push(attr))
```
@nick29581 Is this implementation of the `fold_annotatable` macro OK?

nick29581 commented:
@bors r+ e7f010ef76ebc3fc0fc8261147236c7cbf9a1048

bors commented:
:hourglass: Testing commit e7f010e with merge 31c6018...

bors commented:
:broken_heart: Test failed - [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/3277)

aochagavia commented:
Looks like the pretty printer didn't handle attributes on typedefs.

```
impl Trait for Type {
    #[something]
    type AssocType = AnotherType;
}
```

was printed as

```
impl Trait for Type {type
    AssocType
    =
    AnotherType;
}
```

I have just fixed it. r? @nick29581

nick29581 commented:
@aochagavia r+, but needs a rebase

aochagavia commented:
After rebasing I am getting the following error (in different places):
```
C:\Users\...\Desktop\rust\src\libsyntax\ext\base.rs:33:3: 33:47 error: incorrect stability attribute type
C:\Users\...\Desktop\rust\src\libsyntax\ext\base.rs:33 #[deprecated="Replaced by MultiItemDecorator"]
```

I tried adding `#[unstable]` but it doesn't fix it:

```
C:\Users\...\Desktop\rust\src\libsyntax\ext\base.rs:32:3: 32:11 error: incorrect stability attribute type
C:\Users\...\Desktop\rust\src\libsyntax\ext\base.rs:32 #[unstable]
                                                         ^~~~~~~~
C:\Users\...\Desktop\rust\src\libsyntax\ext\base.rs:33:3: 33:47 error: incorrect stability attribute type
C:\Users\...\Desktop\rust\src\libsyntax\ext\base.rs:33 #[deprecated="Replaced by MultiItemDecorator"]
```

Have you seen this before? What's going on?

nick29581 commented:
@aochagavia probably something to do with http://discuss.rust-lang.org/t/psa-important-info-about-rustcs-new-feature-staging/82 (I haven't read it all myself nor have I seen these errors, so I can't be more specific)

aochagavia commented:
@nick29581 Thanks for the tip! It should be all right now.

nick29581 commented:
@bors r+ 143f2db3174103e459218958f567985b1f47944b

bors commented:
:hourglass: Testing commit 143f2db with merge ec03896...

bors commented:
:broken_heart: Test failed - [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/3413)

aochagavia commented:
@nick29581 I am unable to find the source of the failure. The output of `--pretty expanded` is really weird and I just can't make sense of it. Maybe this issue is too much for me.

nick29581 commented:
@aochagavia A good start would be to reduce the test case to just a single expansion (it currently tests all the possible expansions, which will generate a lot of expanded code). You should then be able to tell whether there is one particular expansion which is broken or whether all of them have the same bug. If you post the output of `--pretty expanded` here (for the broken expansion if only one or any one of them if all of them) then I can try to help you make sense of it.

Manishearth commented:
Needs rebase.

aochagavia commented:
@Manishearth, @nick29581  I'm not going to finish this PR, since I would need to spend too much time to figure out how to fix it.

Manishearth commented:
(I might take this over)

nick29581 commented:
@Manishearth let me know if you decide either way, if you don't, I'll try and take it over the line

Manishearth commented:
Already started. 

https://github.com/Manishearth/rust/compare/multimodifier

Manishearth commented:
I'm getting this on the test, without pretty:

```
/opt/rust/src/test/run-pass-fulldeps/macro-crate-multi-decorator.rs:50:13: 50:19 error: `MyCopy` does not name a structure
/opt/rust/src/test/run-pass-fulldeps/macro-crate-multi-decorator.rs:50     let s = MyCopy { number: 42 };
                                                                                   ^~~~~~
```

I'll need to have a closer look; will do later today.

Manishearth commented:
Okay, can repro the failure

Manishearth commented:
@nick29581 I won't have time to work on this for the next few days, feel free to take over :)

nick29581 commented:
@Manishearth I'm in no rush. If you think you'll come back to it, it's probably better to keep it. I have a lot on too...

Manishearth commented:
Not sure if I will :P

nrc commented:
Finished in #25024

aochagavia commented:
Awesome!
